x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java

```diff
@@ -62,63 +85,159 @@ public interface SecurityMigration {
+                }
+            }, listener::onFailure));
+        }
+
+        private void updateRolesByQuery(
+            SecurityIndexManager indexManager,
+            Client client,
+            BoolQueryBuilder filterQuery,
+            ActionListener<Void> listener
+        ) {
+            UpdateByQueryRequest updateByQueryRequest = new UpdateByQueryRequest(indexManager.getConcreteIndexName());
+            updateByQueryRequest.setQuery(filterQuery);
+            updateByQueryRequest.setScript(
+                new Script(ScriptType.INLINE, "painless", "ctx._source.metadata_flattened = ctx._source.metadata", Collections.emptyMap())
+            );
+            client.admin()
+                .cluster()
+                .execute(UpdateByQueryAction.INSTANCE, updateByQueryRequest, ActionListener.wrap(bulkByScrollResponse -> {
+                    logger.info("Migrated [" + bulkByScrollResponse.getTotal() + "] roles");
+                    listener.onResponse(null);
+                }, listener::onFailure));
+        }
+
+        @Override
+        public Set<NodeFeature> nodeFeaturesRequired() {
+            return Set.of(SecuritySystemIndices.SECURITY_ROLES_METADATA_FLATTENED);
+        }
+
+        @Override
+        public int minMappingVersion() {
+            return ADD_REMOTE_CLUSTER_AND_DESCRIPTION_FIELDS.id();
+        }
+    }
+
+    public static class CleanupRoleMappingDuplicatesMigration implements SecurityMigration {
+        @Override
+        public void migrate(SecurityIndexManager indexManager, Client client, ActionListener<Void> listener) {
+            getRoleMappings(client, ActionListener.wrap(roleMappings -> {
+                List<String> roleMappingsToDelete = getDuplicateRoleMappingNames(roleMappings);
+                if (roleMappingsToDelete.isEmpty() == false) {
+                    logger.info("Found [" + roleMappingsToDelete.size() + "] role mappings to cleanup in .security index.");
+                    deleteNativeRoleMappings(client, roleMappingsToDelete.iterator(), listener);
+                } else {
+                    listener.onResponse(null);
+                }
+            }, listener::onFailure));
+        }
+
+        private void getRoleMappings(Client client, ActionListener<ExpressionRoleMapping[]> listener) {
+            executeAsyncWithOrigin(
+                client,
+                SECURITY_ORIGIN,
+                GetRoleMappingsAction.INSTANCE,
+                new GetRoleMappingsRequestBuilder(client).request(),
+                ActionListener.wrap(response -> listener.onResponse(response.mappings()), listener::onFailure)
+            );
+        }
+
+        private void deleteNativeRoleMappings(Client client, Iterator<String> namesIterator, ActionListener<Void> listener) {
+            String name = namesIterator.next();
+            executeAsyncWithOrigin(
+                client,
+                SECURITY_ORIGIN,
+                DeleteRoleMappingAction.INSTANCE,
```

Comment:
WDYT about a bulk-delete against the raw index instead? Using the API is nice but it's not one-shot where if a single delete OP fails the whole migration stops (at least based on the current implementation). I haven't checked yet but the Header warnings the delete API returns due to name-clash might get logged too, which could confuse users.

Reply:
Yes, that would be ideal and I actually went down that route first, since it's really nice to be able to do it in one shot. 

The reason I switched to use the action was the cache clearing [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java#L215). Just updating the index won't clear the cache so we'd have to add a step to that too in the migration. 

Adding a step to clear the cache might be a good option, but I opted for doing it one by one instead to make sure we use a well tested path (it's another thing that could fail). 

>I haven't checked yet but the Header warnings the delete API returns due to name-clash might get logged too, which could confuse users.

Since we use the transport action directly (no http headers), they should not get logged. I can do some testing to confirm.

Reply:
[This class](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/AbstractRoleMapperClearRealmCache.java) is responsible for cache clearing and all it does is call a transport action with the realm names. 

We could do that and skip the names since that means a clear all realms. Looks like we also need to clear the local cache realms. All of these things require pretty rigorous testing and it's easy to mess up. 

We've decided to instead make sure that if a delete fails, the following ones are executed, to make sure one failed delete doesn't cause all of them to fail.